### PR TITLE
Handle potential coroutine leak issue

### PR DESCRIPTION
**Pull request by wxing1292**

### CODE REVIEW

File: service/history/timerGate.go [hunk 1]
```diff
@@ -104,8 +104,11 @@ func NewLocalTimerGate() LocalTimerGate {
 		for {
 			select {
 			case <-timer.timer.C:
+				select {
 				// re-transmit on gateC
-				timer.fireChan <- struct{}{}
+				case timer.fireChan <- struct{}{}:
+				default:
```
Comment (dmetzgar) [inline service/history/timerGate.go]: Still new to go. Will this prevent a send to the `fireChan` channel if there is already something in the channel?

File: service/history/timerGate.go [hunk 2]
```diff
@@ -88,7 +88,7 @@ func NewLocalTimerGate() LocalTimerGate {
 	timer := &LocalTimerGateImpl{
 		timer:          time.NewTimer(0),
 		nextWakeupTime: time.Time{},
-		fireChan:       make(chan struct{}),
+		fireChan:       make(chan struct{}, 1),
```
Comment (vancexu) [inline service/history/timerGate.go]: Will this cause skipping timer in fireChan?

Comment (wxing1292) [reply service/history/timerGate.go]: nop, the code, after the change looks like this
```
				select {
				// re-transmit on gateC				
				case timer.fireChan <- struct{}{}:
				default:
				}
```

Comment (wxing1292) [reply service/history/timerGate.go]: so if there is a signal in the channel already, just skip

File: service/history/timerGate.go (hunk 1 above)
Comment (dmetzgar) [reply service/history/timerGate.go]: Looks like you already answered this.

Comment (wxing1292) [reply service/history/timerGate.go]: yes, if the channel is full, then it means there is already a signal inside.

this class aims to send out signal after given time, (like a timer). if the receiving side of the channel is not polling a message in the channel, adding more does not make anything better.

**Pull request merged**